### DISCUSSION TITLE
How to integrate with a crate that uses `main` (binaries)?

### DISCUSSION THREAD

**Discussion by milesj** [napi-rs/napi-rs#1077] [Q&A]
I have a project named moon that provides a CLI written in Rust (using clap), and I want it to be installed through npm. Ideally this CLI is executed when `yarn moon`, `npm moon`, or `npx moon`, etc, is called.

I'm well aware of napi-rs working for _libraries_, but I'm completely lost on if it's possible for _binaries_. I have an open PR where I'm trying things: https://github.com/milesj/moon/pull/56. Here's also a bit more context and direct links.

- The cli crate without napi integration https://github.com/milesj/moon/tree/release/crates/cli (I want this to be the build)
- The core crate with napi integration https://github.com/milesj/moon/tree/release/crates/core (it's empty for now as im trying to get the lib build working)
- The cli npm package https://github.com/milesj/moon/blob/release/packages/cli/package.json (see triples packages up a directory)
- And the napi build scripts https://github.com/milesj/moon/blob/release/package.json#L14

I haven't seen anything in the docs about this, so looking for guidance... if it's even possible.

Also, on a side note, I could not get js/dts generate working.

**Comment by milesj** [comment]
I'm looking at `parcel-css` which does exactly what I'm asking (as they use clap), and it looks like they just built it with cargo. Maybe I don't need n-api for this? But it would be nice if this _was supported_.

**Reply by Brooooooklyn** [reply]
@milesj If you are building some executable binaries, and they are not expected to be required in `JavaScript` as a library, just build it via `Cargo`.

**Reply by milesj** [reply]
@Brooooooklyn Yeah I eventually figured that out. It would be nice if n-api CLI supported binaries, since the rest of the tooling (artifacts, CI actions, etc) are still beneficial.

**Reply by messense** [reply]
#1107
